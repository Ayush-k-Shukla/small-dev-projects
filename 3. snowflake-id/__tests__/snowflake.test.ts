import SnowFlake from '../src';

describe('Snowflake ID Generator', () => {
  let generator: SnowFlake;

  beforeEach(() => {
    generator = new SnowFlake(1);
  });

  test('Generated IDs should be unique', () => {
    const ids = new Set();
    for (let i = 0; i < 1000; i++) {
      const id = generator.generateId();
      expect(ids.has(id)).toBe(false);
      ids.add(id);
    }
  });

  test('Generated IDs should be in increasing order', () => {
    let prevId = generator.generateId();
    for (let i = 0; i < 1000; i++) {
      const newId = generator.generateId();
      expect(newId > prevId).toBe(true);
      prevId = newId;
    }
  });

  test('Machine ID out of range should throw an error', () => {
    expect(() => new SnowFlake(1024)).toThrowErrorMatchingSnapshot();
    expect(() => new SnowFlake(-1)).toThrowErrorMatchingSnapshot();
  });

  test('Clock moving backwards should throw an error', () => {
    generator['lastTimeStamp'] = generator['currentTimeStamp']() + 1n;
    expect(() => generator.generateId()).toThrowErrorMatchingSnapshot(
      'Clock moved backwards. Refusing to generate ID'
    );
  });

  test('Timestamp should be within 41 bits', () => {
    const id = generator.generateId();
    const timestamp = id >> (10n + 12n);
    expect(timestamp < 1n << 41n).toBe(true);
  });
});
