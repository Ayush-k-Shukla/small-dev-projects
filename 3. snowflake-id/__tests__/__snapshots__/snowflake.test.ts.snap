// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snowflake ID Generator Clock moving backwards should throw an error: Clock moved backwards. Refusing to generate ID 1`] = `"Clock moved backwards. Refusing to generate ID"`;

exports[`Snowflake ID Generator Machine ID out of range should throw an error 1`] = `"Machine ID must be between 0 and 1023"`;

exports[`Snowflake ID Generator Machine ID out of range should throw an error 2`] = `"Machine ID must be between 0 and 1023"`;
